lib/companies/schema/company.ex

```diff
@@ -24,8 +26,23 @@ defmodule Companies.Schema.Company do
 
   @doc false
   def changeset(company, attrs) do
+    attrs = Map.merge(attrs, slug_map(attrs))
+
     company
-    |> cast(attrs, [:name, :description, :url, :github, :location, :blog, :industry_id])
+    |> cast(attrs, [:name, :description, :url, :github, :location, :blog, :industry_id, :slug])
     |> validate_required([:name, :description, :url, :industry_id])
```

Comment:
The slug is unique in the database by a constraint but this constraint is not handled in the changeset. 

Another thing to deal with is how pending changes are managed in the code. When a company is created a pending_change is inserted in the database. When that change is approved, the company will be inserted in the companies table and could fail on the uniqueness constraint. 

It would be preferable imo to deal with duplicate slugs earlier, perhaps by adding an [unsafe_validate_unique](https://hexdocs.pm/ecto/Ecto.Changeset.html#unsafe_validate_unique/3) and adding postfixes to the slugs to make them unique.

This would not entirely be foolproof (e.g. multiple pending changes could have the same slug) but could go a long way.

Reply:
we could add the company id in the slug just to be safe. e.g. Plataformatec turns into 1-plataformatec

Reply:
Crazy idea feel free to shoot this down: what if we _don't_ enforce the uniqueness in the changeset (or we create a changeset without it) and instead defer the work of creating unique slugs onto admins approving changes. In the UI to approve a company we could show whether the slug was already taken and require the admin to make a unique one.

Reply:
@maartenvanvliet / @gemantzu how do you both feel about moving generation of things like slugs (or password hashes) into the changeset pipeline like such?

```elixir
def changeset(company, attrs) do
  company
  |> cast(attrs, [:name, :description, :url, :github, :location, :blog, :industry_id, :slug])
  |> validate_required([:name, :description, :url, :industry_id])
  |> generate_slug()
  |> unique_constraint(:slug)
end

# when there is an existing non-nil `id` in the data, we don't want to regenerate a slug even on name change
# that would break any potential backlinks.
defp generate_slug(%Ecto.Changeset{data: %{id: id}} = changeset) when not is_nil(id) do
  changeset
end

defp generate_slug(%Ecto.Changeset{changes: %{name: name}} = changeset) do
  slug =
    name
    |> String.replace(~r/['’]s/u, "s")
    |> String.downcase()
    |> String.replace(~r/([^a-z0-9가-힣])+/, "-")
    |> String.replace(" ", "-")

  put_change(changeset, :slug, slug)
end
```